src/components/spatial/SpatialSubscriber.js

```diff
@@ -14,162 +16,131 @@ import {
   CLEAR_PLEASE_WAIT,
   VIEW_INFO,
   CELL_SETS_VIEW,
-  RASTER_ADD,
-  SLIDERS_CHANGE,
-  COLORS_CHANGE,
-  CHANNEL_VISIBILITY_CHANGE,
+  LAYER_ADD,
+  LAYER_REMOVE,
+  LAYER_CHANGE,
 } from '../../events';
 import Spatial from './Spatial';
 
-export default class SpatialSubscriber extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      cells: {},
-      selectedCellIds: new Set(),
-      cellColors: null,
-      colorValues: null,
-      sliderValues: null,
-      channelVisibilities: null,
-      raster: null,
+export default function SpatialSubscriber({
+  children,
+  onReady,
+  removeGridComponent,
+  moleculeRadius,
+  view,
+  cellRadius,
+  uuid = null,
+}) {
+  const [cells, setCells] = useState(null);
+  const [molecules, setMolecules] = useState(null);
+  const [cellColors, setCellColors] = useState(null);
+  const [neighborhoods, setNeighborhoods] = useState(null);
+  const [selectedCellIds, setSelectedCellIds] = useState(new Set());
+  const [imageLayerProps, setImageLayerProps] = useState({});
+  const [imageLayerLoaders, setImageLayerLoaders] = useState({});
+
+  const onReadyCallback = useCallback(onReady, []);
+
+  useEffect(() => {
+    const moleculesAddSubscriber = (msg, newMolecules) => setMolecules(newMolecules);
+    const nbhdsAddSubscriber = (msg, newNeighborhoods) => setNeighborhoods(newNeighborhoods);
+    const cellsAddSubscriber = (msg, newCells) => setCells(newCells);
+    const cellsSelectionSubscriber = (msg, newCellIds) => setSelectedCellIds(newCellIds);
+    const cellsColorSubscriber = (msg, newColors) => setCellColors(newColors);
+    function layerAddSubscriber(msg, { layerId, loader, layerProps }) {
+      setImageLayerProps(prevLayerProps => ({ ...prevLayerProps, [layerId]: layerProps }));
+      setImageLayerLoaders(prevLoaders => ({ ...prevLoaders, [layerId]: loader }));
+    }
+    function layerChangeSubscriber(msg, { layerId, layerProps }) {
+      setImageLayerProps(prevLayerProps => ({
+        ...prevLayerProps,
+        [layerId]: { ...prevLayerProps[layerId], ...layerProps },
+      }));
+    }
+    function layerRemoveSubscriber(msg, layerId) {
+      setImageLayerLoaders((prevLoaders) => {
+        const { [layerId]: _, ...nextLoaders } = prevLoaders;
+        return nextLoaders;
+      });
+      setImageLayerProps((prevLayerProps) => {
+        const { [layerId]: _, ...nextLayerProps } = prevLayerProps;
+        return nextLayerProps;
+      });
+    }
+
+    const moleculesAddToken = PubSub.subscribe(MOLECULES_ADD, moleculesAddSubscriber);
+    const nbhdsAddToken = PubSub.subscribe(NEIGHBORHOODS_ADD, nbhdsAddSubscriber);
```

Comment:
I'd spell it out. (Down below, I couldn't guess what it meant. :) )

Reply:
Sorry it was to avoid the eslint line length rule. Perhaps I'll ignore that for this block.